### PR TITLE
Amazing Performance intro post

### PR DESCRIPTION
**Pull request by RobbieMcKinstry**
This post is the first in a series about the Amazing Performance initiative. It introduces the tools we've built in the past to help us analyze the perf of Pulumi programs, and describes two small wins as part of the initiative.

### CODE REVIEW

**Comment by susanev** [comment]
@RobbieMcKinstry the appdash screenshot is a mb, does it need to be that big? mind if i push a commit making the file size smaller? as an additional note our blog cannot display anything with a larger width than 900px so there is no need for images to have a greater width than 900.

**Comment by RobbieMcKinstry** [comment]
@susanev Thanks very much for offering! I'll take you up on that! That sounds good to me :)

File: themes/default/content/blog/amazing-performance/index.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+Our more experienced TypeScript users also ran up against this issue. One pattern used in CI environments or with [Automation API](https://www.pulumi.com/docs/guides/automation-api/) is to typecheck Pulumi programs separately from execution. Some users enforce typechecking during pull requests to guarantee that all code hitting the repository's main branch is valid. Once it's typechecked, they can precompile the TypeScript into JavaScript before execution. Finally, when it comes time to execute their Pulumi program, they run the compiled JavaScript, which they know to be well-typed. The motivation for precompiling comes from using your own preprocessor. Some customers want to ditch TS-Node for compilation and prefer to use [SWC](https://swc.rs/), [ESBuild](https://esbuild.github.io/), or other options for increased performance or consistency with the rest of their codebase.
+
+In both scenarios, importing typescript when it's not used results in a 300ms slowdown. In [a small PR](https://github.com/pulumi/pulumi/pull/10214), we detect cases that don't require TypeScript and dynamically import it only when needed.
```
Comment (kpitzen) [inline themes/default/content/blog/amazing-performance/index.md:72]:
```suggestion
In both scenarios, importing TypeScript when it's not used results in a 300ms slowdown. In [a small PR](https://github.com/pulumi/pulumi/pull/10214), we detect cases that don't require TypeScript and dynamically import it only when needed.
```

**Review by kpitzen — APPROVED** [review]
My tiniest nit yet - otherwise this looks fantastic!

File: themes/default/content/blog/amazing-performance/index.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+authors:
+    - robbie-mckinstry
+
+tags:
+    - performance
+    - platform
+    - engineering
+---
+
+This is the first post in a series about performance optimizations we've made to the Pulumi CLI. Over the last six months at Pulumi, the Platform Team has been working on a project we call "Amazing Performance." Amazing Performance is a new initiative to improve the throughput and latency of the Pulumi CLI not only for power users but for everyone. By the end of June 2022, we assembled a list of issues containing both high-value improvements requiring a sizable investment and low-hanging fruit for quick wins. The full list, including the items we have yet to tackle, is contained in [a tracking issue on GitHub](https://github.com/pulumi/home/issues/1499). This blog series will cover the highlights.
```
Comment (iwahbe) [inline themes/default/content/blog/amazing-performance/index.md:16]: The tracking issue is private within pulumi. I don't think we should link to it.

**Comment by susanev** [comment]
@RobbieMcKinstry did you want to merge this soon? can i help in any way?

**Comment by RobbieMcKinstry** [comment]
@susanev I've been moving on this slowly for no particular reason. I need to address @iwahbe's comment and do a final editing pass.

File: themes/default/content/blog/amazing-performance/index.md (hunk 2 above)
Comment (RobbieMcKinstry) [reply themes/default/content/blog/amazing-performance/index.md:16]: @t0yv0 also pointed this out. You're right, we shouldn't link to issues in Home.
Maybe about a month back, I talked with @mikhailshilkov about whether we could exfiltrate the issue into the public space. I opened https://github.com/pulumi/pulumi/issues/11598, which is essentially a forward-looking version of the private issue, which we intend to close as it was specifically scoped to Q4 2022.

File: themes/default/content/blog/amazing-performance/index.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
+---
+title: "Amazing Performance"
```
Comment (gthuang) [inline themes/default/content/blog/amazing-performance/index.md:2]: i suggest coming up with a more descriptive title. i like "amazing performance", so would recommend expanding this. perhaps something like "Achieving Amazing Performance for the Pulumi CLI"?

File: themes/default/content/blog/amazing-performance/index.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+authors:
+    - robbie-mckinstry
+
+tags:
+    - performance
+    - platform
+    - engineering
+---
+
+This is the first post in a series about performance optimizations we've made to the Pulumi CLI. Over the last six months at Pulumi, the Platform Team has been working on a project we call "Amazing Performance." Amazing Performance is a new initiative to improve the throughput and latency of the Pulumi CLI not only for power users but for everyone. By the end of June 2022, we assembled a list of issues containing both high-value improvements requiring a sizable investment and low-hanging fruit for quick wins. The full list, including the items we have yet to tackle, is contained in [a tracking issue on GitHub](https://github.com/pulumi/pulumi/issues/11598). This blog series will cover the highlights.
```
Comment (gthuang) [inline themes/default/content/blog/amazing-performance/index.md:16]: do folks know what Platform Team means at pulumi?

File: themes/default/content/blog/amazing-performance/index.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+<!--more-->
+
+This post has two sections. First, we'll describe the tools we built previously to track performance. Secondly, we'll recap a few small performance wins. Future posts will detail our major wins as part of the Amazing Performance initiative.
+
+## Tools We've Already Built
```
Comment (gthuang) [inline themes/default/content/blog/amazing-performance/index.md:22]: a more apt title might be something like "How we measure and track performance"

File: themes/default/content/blog/amazing-performance/index.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+Pulumi can send traces to a Zipkin server, or aggregate them to a local file and let you view them offline. If you provide an HTTP URI to the `--tracing` flag, then Pulumi assumes that URI points to a web server that accepts Zipkin-format traces, and it will send all tracing data there. If you provide a file URI instead, Pulumi spins up a local Zipkin server to which child processes send their traces. Once execution is complete, Pulumi aggregates the traces into a single local file and spins down the Zipkin server. You can run `PULUMI_DEBUG_COMMANDS=1 pulumi view-trace ./up.trace` to view the trace in an embedded web UI.
+
+![An image of a Pulumi trace](appdash.jpg)
+
+Not all spans are exceptionally well-named, and there are blind spots in our traces. However, traces are typically a reliable way to subdivide the program execution, so hotspots are easily identifiable. In the future, we plan to migrate to OpenTelemetry and improve our trace coverage. (Feel free to get in touch on the [Community Slack](https://slack.pulumi.com/) if you want to help with this effort!)
+
+## Fixing Low Hanging Fruit
```
Comment (gthuang) [inline themes/default/content/blog/amazing-performance/index.md:60]:
```suggestion
## Fixing Low-Hanging Fruit to Increase Speed
```

**Review by gthuang — COMMENTED** [review]
nice blog! just a few suggestions.

**Comment by aaronkao** [comment]
Quick thing, please don't merge this blog post until you chat with @scottslowe , @cnunciato , or myself. We are trying to space out the blog post so there is a nice cadence and flow. Thanks

**Comment by RobbieMcKinstry** [comment]
Going through and addressing @gthuang's feedback presently!

File: themes/default/content/blog/amazing-performance/index.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    - performance
+    - platform
+    - engineering
+---
+
+This is the first post in a series about performance optimizations we've made to the Pulumi CLI. Over the last six months at Pulumi, the Platform Team has been working on a project we call "Amazing Performance." Amazing Performance is a new initiative to improve the throughput and latency of the Pulumi CLI not only for power users but for everyone. By the end of June 2022, we assembled a list of issues containing both high-value improvements requiring a sizable investment and low-hanging fruit for quick wins. The full list, including the items we have yet to tackle, is contained in [a tracking issue on GitHub](https://github.com/pulumi/pulumi/issues/11598). This blog series will cover the highlights.
+
+<!--more-->
+
+This post has two sections. First, we'll describe the tools we built previously to track performance. Secondly, we'll recap a few small performance wins. Future posts will detail our major wins as part of the Amazing Performance initiative.
```
[the remark this answers is not in the document]
Comment (RobbieMcKinstry) [reply themes/default/content/blog/amazing-performance/index.md:20]: I switched the wording to "quick wins". I feel like that better serves my intent, which is to say these PRs aren't so big as to justify a longer explainer, and as a result the performance gain will be smaller.

File: themes/default/content/blog/amazing-performance/index.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+Not all spans are exceptionally well-named, and there are blind spots in our traces. However, traces are typically a reliable way to subdivide the program execution, so hotspots are easily identifiable. In the future, we plan to migrate to OpenTelemetry and improve our trace coverage. (Feel free to get in touch on the [Community Slack](https://slack.pulumi.com/) if you want to help with this effort!)
+
+## Fixing Low Hanging Fruit
+
+Some of the work we completed for Amazing Performance targeted simple, isolated changes to make the system a little snappier. The rest of this blog details two quick examples.
+
+### Skipping a Slow Import
+
+When Pulumi executes NodeJS code, it compiles TypeScript into JavaScript using [TS-Node](https://www.npmjs.com/package/ts-node). Consequently, Pulumi bundles the TypeScript compiler into its NodeJS runtime so TS-Node can perform the compilation.
```
[the remark this answers is not in the document]
Comment (RobbieMcKinstry) [reply themes/default/content/blog/amazing-performance/index.md:66]: Done, both here and below.

**Comment by RobbieMcKinstry** [comment]
Feedback addressed! This guy is ready to go!

File: themes/default/content/blog/amazing-performance/index.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
+---
+title: "Achieving Amazing Performance in the Pulumi CLI"
+date: 2023-01-27T07:00:00-05:00
```
Comment (cnunciato) [inline themes/default/content/blog/amazing-performance/index.md:3]:
```suggestion
date: 2023-01-26
```

File: themes/default/content/blog/amazing-performance/index.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+Users typically hit this bug in one of two ways. Some users choose to write Pulumi programs in pure JavaScript. Pulumi was needlessly importing the TypeScript compiler, even when there was nothing to compile.
+
+The second way a user might encounter this bug in a more advanced deployment pipeline. One common advanced pattern in CI environments or with [Automation API](https://www.pulumi.com/docs/guides/automation-api/) is to typecheck Pulumi programs separately from execution. CI will enforce typechecking during pull requests to guarantee that all code hitting the repository's main branch is valid. Once it's typechecked and merged, the Pulumi program can precompile the TypeScript into JavaScript before execution. Finally, when it comes time to execute the Pulumi program, the deployment pipeline will run the compiled JavaScript, which is knowned to be well-typed. The motivation for precompiling comes from using your own preprocessor. Some customers want to ditch TS-Node for compilation and prefer to use [SWC](https://swc.rs/), [ESBuild](https://esbuild.github.io/), or other options for increased performance, or for consistency with the rest of their codebase.
```
Comment (cnunciato) [inline themes/default/content/blog/amazing-performance/index.md:72]:
```suggestion
The second way a user might encounter this bug is in a more advanced deployment pipeline. One common advanced pattern in CI environments or with [Automation API](https://www.pulumi.com/docs/guides/automation-api/) is to typecheck Pulumi programs separately from execution. CI will enforce typechecking during pull requests to guarantee that all code hitting the repository's main branch is valid. Once it's typechecked and merged, the Pulumi program can precompile the TypeScript into JavaScript before execution. Finally, when it comes time to execute the Pulumi program, the deployment pipeline will run the compiled JavaScript, which is known to be well-typed. The motivation for precompiling comes from using your own preprocessor. Some customers want to ditch TS-Node for compilation and prefer to use [SWC](https://swc.rs/), [ESBuild](https://esbuild.github.io/), or other options for increased performance, or for consistency with the rest of their codebase.
```

**Review by cnunciato — COMMENTED** [review]
Couple little suggestions and a meta image and then I think we're good to go.

**Review by cnunciato — APPROVED** [review]

**Pull request merged**